DOCS-11608-whatsNewDW-duke (#13)
mulesoft/docs-dataweave: modules/ROOT/pages/whats-new-in-dw.adoc

= What's New in DataWeave

The 2.4 version of DataWeave introduces the following features:

> * Ability to read larger-than-memory strings automatically. When you are using the indexed reader strategy and processing a String with a size larger than 1.5 MB, DataWeave automatically splits the value in chunks to avoid out-of-memory issues. Related reader properties `com.mulesoft.dw.max_memory_allocation` and `com.mulesoft.dw.buffered_char_sequence.enabled`. See xref:dataweave-formats.adoc[Supported Data Formats].

Reviewer: ```suggestion
* Ability to read larger-than-memory strings automatically. When you are using the indexed reader strategy and processing a String with a size larger than 1.5 MB, DataWeave automatically splits the value in chunks to avoid out-of-memory issues. Related reader properties `com.mulesoft.dw.max_memory_allocation` and `com.mulesoft.dw.buffered_char_sequence.enabled`. Refer to xref:dataweave-formats.adoc[Supported Data Formats].
```

Reviewer: Just for accessibility consideration

* New reader and writer properties for working with data formats. For example, the XML writer now enables you to define namespaces at the root level. 
* Many new modules, functions, types, annotations, and variables. Some are experimental and subject to change in future releases.
* A number of functions are overloaded to introduce new helper functions for

> handling `null` values.

Reviewer: ```suggestion
managing `null` values.
```

Reviewer: Per Kim suggestion

Author: I think in this specific context, it's more common to say "handling."

Changes to the documentation include more detailed function signatures and organization of the functions by their module names, such as `dw::Core`, `dw::core::Strings`, and `dw::util::Math`.

> DataWeave 2.4 is bundled with Mule 4.4. For information about the Mule 4.4 release, see xref:4.4@mule-runtime::whats-new-in-mule.adoc[What's New in Mule 4.4].

Reviewer: ```suggestion
DataWeave 2.4 is bundled with Mule 4.4. For information about the Mule 4.4 release, refer to xref:4.4@mule-runtime::whats-new-in-mule.adoc[What's New in Mule 4.4].
```

= What's New in DataWeave

The 2.4 version of DataWeave introduces the following features:

* Larger than memory strings.
* New reader and writer properties for working with data formats.
* Many new modules, functions, types, annotations, and variables. Some are experimental and subject to change in future releases.
* A number of other overloaded functions introduce new helper functions for
handling `null` values.

Changes to the documentation include more detailed function signatures and organization of the functions by their module names, such as `dw::Core`, `dw::core::Strings`, and `dw::util::Math`.

DataWeave 2.4 is bundled with Mule 4.4. For information about the Mule 4.4 release, see xref:4.4@mule-runtime::whats-new-in-mule.adoc[What's New in Mule 4.4].

== New Data Formats and Properties

DataWeave introduces the following reader and writer properties:

* DataWeave (dw) reader property: `privileges`

> * Excel (XLSX) reader property: `tableLimit`

Author: ```suggestion
* Excel (XLSX) reader property: `tableLimit` enables you to set position of the last column in each row. 
```

> * YAML writer property: `writeDeclaration`

Author: ```suggestion
* YAML writer property: `writeDeclaration` indicates whether to write the header declaration or not.
```

Reviewer 2: LGTM

> * XML writer properties: `defaultNamespace`, `writeDeclaredNamespaces`

Author: ```suggestion
* XML writer properties: `defaultNamespace` specifies the default namespaces of the output XML , and `writeDeclaredNamespaces` marks the namespaces to declare in the root element of the XML.
```

Reviewer 2: LGTM

* xref:dw-strings.adoc[dw::core::Strings] introduces new functions:
  ** xref:dw-strings-functions-collapse.adoc[collapse]
  ** xref:dw-strings-functions-countcharactersby.adoc[countCharactersBy]
  ** xref:dw-strings-functions-countmatches.adoc[countMatches]
  ** xref:dw-strings-functions-everycharacter.adoc[everyCharacter]
  ** xref:dw-strings-functions-first.adoc[first]
  ** xref:dw-strings-functions-hammingdistance.adoc[hammingDistance]
  ** xref:dw-strings-functions-last.adoc[last]
  ** xref:dw-strings-functions-levenshteindistance.adoc[levenshteinDistance]
  ** xref:dw-strings-functions-lines.adoc[lines]
  ** xref:dw-strings-functions-mapstring.adoc[mapString]
  ** xref:dw-strings-functions-remove.adoc[remove]
  ** xref:dw-strings-functions-replaceall.adoc[replaceAll]
  ** xref:dw-strings-functions-reverse.adoc[reverse]
  ** xref:dw-strings-functions-somecharacter.adoc[someCharacter]
  ** xref:dw-strings-functions-substring.adoc[substring]
  ** xref:dw-strings-functions-substringby.adoc[substringBy]
  ** xref:dw-strings-functions-substringevery.adoc[substringEvery]
  ** xref:dw-strings-functions-words.adoc[words]

> * xref:dw-runtime-types.adoc[Runtime Types] introduces new types, including `MimeType`, `Location`, `Position`, `LogLevel`, `LogEntry`, and `DataFormatProperty`.

Author: Removed experimental Runtime functions. Listed non-experimental Runtime types that are new.

* xref:dw-coercions.adoc[dw::util::Coercions] is a new DataWeave module with functions and types.
* xref:dw-math.adoc[dw::util::Math] is a new DataWeave module with functions and variables.
* xref:dw-periods.adoc[dw::util::Periods] is a new DataWeave module with functions.

> ** xref:dw-tree.adoc[dw::util::Tree] introduces new functions:

Reviewer 2: wrong indentation, should be: 
* xref:dw-tree.adoc[dw::util::Tree] introduces new functions:

Author: ```suggestion
* xref:dw-tree.adoc[dw::util::Tree] introduces new functions:
```

Author: fixed indentation per Lische's comment

= What's New in DataWeave

The 2.4 version of DataWeave introduces the following features:

* Larger than memory strings.
* New reader and writer properties for working with data formats.
* Many new modules, functions, types, annotations, and variables. Some are experimental and subject to change in future releases.
* A number of other overloaded functions introduce new helper functions for
handling `null` values.

Changes to the documentation include more detailed function signatures and organization of the functions by their module names, such as `dw::Core`, `dw::core::Strings`, and dw::util::Math.

DataWeave 2.4 is bundled with Mule 4.4. For information about the Mule 4.4 release, see xref:4.4@mule-runtime::whats-new-in-mule.adoc[What's New in Mule 4.4].

== New Data Formats and Properties

> 

Author: ```suggestion

The following reader and writer properties for DataWeave input and output are new:
```

* DataWeave (dw) reader property: `privileges`
* Excel (XLSX) reader property: `tableLimit`
* YAML writer property: `writeDeclaration`

> * XML writer property: `defaultNamespace`

Author: ```suggestion
* XML writer properties: `defaultNamespace`, `writeDeclaredNamespaces`
```

= What's New in DataWeave

The 2.4 version of DataWeave introduces the following features:

* Larger than memory strings.

> * *TODO: Support for the binary format. IS THIS THE DWB FORMAT?*

Author: removing per PM & mngt

= What's New in DataWeave

The 2.4 version of DataWeave introduces the following features:

> * Larger than memory strings.

Reviewer 3: Can we point to some of the details here?

Author: TODO: yes, that doc is in progress. I'll link to it once I know the file name. https://www.mulesoft.org/jira/browse/DOCS-12018

Author: ```suggestion
* Ability to read larger-than-memory strings by setting the indexed reader property `com.mulesoft.dw.max_memory_allocation` and optionally, the performance-related property `com.mulesoft.dw.buffered_char_sequence.enabled`. See xref:dataweave-formats.adoc[Supported Data Formats].
```

Reviewer 2: This could be more accurate: 
Ability to read larger-than-memory strings. 
When you are using the indexed reader strategy and processing a `String` with a size larger than 1.5 MB, DataWeave automatically splits the value in chunks to avoid out-of-memory issues. See xref:dataweave-formats.adoc[Supported Data Formats].

Author: @mlischetti Updated the suggestion with your description while keeping the new configuration properties in it. Does this work?

Reviewer 2: Yes, it works. 
One additional note:  com.mulesoft.dw.max_memory_allocation is not a new reader properties. 
So we can: 
1- Remove the whole "Reader reader properties..." section 
2- Just say, "Related reader properties are `com.mulesoft.dw.max_memory_allocation` and `com.mulesoft.dw.buffered_char_sequence.enabled`"

Author: ok, opted for #2. removed "are new"

> * New reader and writer properties for working with data formats.

Reviewer 3: I think we should explicitly mention that XML now allows defining namespaces at the root level.

Author: ```suggestion
* New reader and writer properties for working with data formats. For example, XML now enables you to define namespaces at the root level. 
```

Reviewer 2: LGMT

* Many new modules, functions, types, annotations, and variables. Some are experimental and subject to change in future releases.

> * A number of other overloaded functions introduce new helper functions for

Author: ```suggestion
* A number of functions are overloaded to introduce new helper functions for
```

Reviewer 2: LGTM

> handling `null` values.

Reviewer 3: This is ok, though I would word it as: a number of functions have been overloaded to handle `null` values. Cause this sounds there were already overloaded.

> handling `null` values.

Author: done, see suggestion above

> Changes to the documentation include more detailed function signatures and organization of the functions by their module names, such as `dw::Core`, `dw::core::Strings`, and dw::util::Math.

Reviewer 3: ```suggestion
Changes to the documentation include more detailed function signatures and organization of the functions by their module names, such as `dw::Core`, `dw::core::Strings`, and `dw::util::Math`.
```

DataWeave 2.4 is bundled with Mule 4.4. For information about the Mule 4.4 release, see xref:4.4@mule-runtime::whats-new-in-mule.adoc[What's New in Mule 4.4].

== New Data Formats and Properties

DataWeave introduces the following reader and writer properties:

> * DataWeave (dw) reader property: `privileges`

Reviewer 3: Should we add a brief description of these?

Author: TODO: privileges

Author: ```suggestion
* DataWeave (dw) reader property: `privileges` enables you to provide a comma-separated list of privileges to use in the format, such as `'Resources,Properties'`.
```

Reviewer 2: LGTM

* Excel (XLSX) reader property: `tableLimit`
* YAML writer property: `writeDeclaration`
* XML writer properties: `defaultNamespace`, `writeDeclaredNamespaces`

== New DataWeave Function Modules and Features

New DataWeave modules and features:

* xref:dw-core.adoc[dw::Core] introduces new functions and annotations:
** xref:dw-core-functions-indexof.adoc[indexOf]
** xref:dw-core-functions-lastindexof.adoc[lastIndexOf]
** xref:dw-core-functions-onnull.adoc[onNull]
** xref:dw-core-functions-then.adoc[then]
** xref:dw-core-annotations.adoc[Core Annotations] introduces a number of new annotations:

> * xref:dw-dates.adoc[dw::core::Dates] is a new module with functions and types.

Reviewer 3: "functions and types to work with dates, such as {example}"

Reviewer 3: the example could be `today`

Author: ```suggestion
* xref:dw-dates.adoc[dw::core::Dates] is a new module with functions and types, for example, the function `today`.
```

Reviewer 2: * xref:dw-dates.adoc[dw::core::Dates] is a new module with functions and types to create and manipulate Date types, for example, the function `today`.

Author: updated comment to suggested and committed it.

* xref:dw-strings.adoc[dw::core::Strings] introduces new functions:
  ** xref:dw-strings-functions-collapse.adoc[collapse]
  ** xref:dw-strings-functions-countcharactersby.adoc[countCharactersBy]
  ** xref:dw-strings-functions-countmatches.adoc[countMatches]
  ** xref:dw-strings-functions-everycharacter.adoc[everyCharacter]
  ** xref:dw-strings-functions-first.adoc[first]
  ** xref:dw-strings-functions-hammingdistance.adoc[hammingDistance]
  ** xref:dw-strings-functions-last.adoc[last]
  ** xref:dw-strings-functions-levenshteindistance.adoc[levenshteinDistance]
  ** xref:dw-strings-functions-lines.adoc[lines]
  ** xref:dw-strings-functions-mapstring.adoc[mapString]
  ** xref:dw-strings-functions-remove.adoc[remove]
  ** xref:dw-strings-functions-replaceall.adoc[replaceAll]
  ** xref:dw-strings-functions-reverse.adoc[reverse]
  ** xref:dw-strings-functions-somecharacter.adoc[someCharacter]
  ** xref:dw-strings-functions-substring.adoc[substring]
  ** xref:dw-strings-functions-substringby.adoc[substringBy]
  ** xref:dw-strings-functions-substringevery.adoc[substringEvery]
  ** xref:dw-strings-functions-words.adoc[words]

> * xref:dw-runtime.adoc[dw::Runtime] introduces new experimental functions:

Reviewer 3: I would not mention experimental functions in this what's new doc, since we do not want to encourage usage of such functions.

  ** xref:dw-runtime-functions-dataformatsdescriptor.adoc[dataFormatsDescriptor]
  ** xref:dw-runtime-functions-eval.adoc[eval]
  ** xref:dw-runtime-functions-evalurl.adoc[evalUrl]
  ** xref:dw-runtime-functions-location.adoc[location]
  ** xref:dw-runtime-functions-run.adoc[run]
  ** xref:dw-runtime-functions-runurl.adoc[runUrl]
	** xref:dw-runtime-types.adoc[Runtime Types] introduces new types.

> * xref:dw-coercions.adoc[dw::util::Coercions] is a new DataWeave module with functions and types.

Reviewer 3: "functions and types to simplify data type transformations, for example `toString`"

Author: ```suggestion
* xref:dw-coercions.adoc[dw::util::Coercions] is a new DataWeave module with functions and types to simplify data type transformations, for example, `toString`.
```

Reviewer 2: LGTM

> * xref:dw-math.adoc[dw::util::Math] is a new DataWeave module with functions and variables.

Reviewer 3: "functions and types covering advanced math, for example `sin`"

Author: it also introduces variables

Author: ```suggestion
* xref:dw-math.adoc[dw::util::Math] is a new DataWeave module with functions and variables, for example, `sin`.
```

Reviewer 2: * xref:dw-math.adoc[dw::util::Math] is a new DataWeave module with functions and variables that provides mathematical functions, for example, the function `sin`.

> * xref:dw-periods.adoc[dw::util::Periods] is a new DataWeave module with functions.

Reviewer 3: "functions and types to simplify the creation of Periods, for example `days`"

Author: ```suggestion
* xref:dw-periods.adoc[dw::util::Periods] is a new DataWeave module with functions and types to simplify the creation of Periods, for example `days`.
```

Reviewer 2: LGTM
